### PR TITLE
Adjust hit location weights & introduce rng to bp selection

### PR DESCRIPTION
**Pull request by macrosblackd**
Tweaking some numbers since monsters seem to be eating faces again abnormally often again.

### CODE REVIEW

File: src/creature.cpp [hunk 1]
```diff
@@ -1587,20 +1588,22 @@ body_part Creature::select_body_part(Creature *source, int hit_roll)
...
-    double roll = rng_float(1, totalWeight);
-    body_part selected_part = bp_torso;
-
-    std::set<std::pair<body_part, double>, weight_compare>::iterator adj_iter;
-    for(adj_iter = adjusted_weights.begin(); adj_iter != adjusted_weights.end(); ++adj_iter) {
-        roll -= adj_iter->second;
-        if(roll <= 0) {
-            selected_part = adj_iter->first;
+    double roll = rng_float(0, totalWeight);
+    body_part selected_part;
```
Comment (kevingranade) [inline src/creature.cpp]: Jenkins is yelling about the chance that this will never be initialized, and I agree, just have it default to bp_torso in case something goes wrong.

File: src/creature.cpp [hunk 2]
```diff
@@ -1587,20 +1588,22 @@ body_part Creature::select_body_part(Creature *source, int hit_roll)
...
+    do {
+        auto adj_iter = random_entry_removed( adjusted_weights );
```
Comment (kevingranade) [inline src/creature.cpp]: We should either use random_entry_bla() OR keep track ourselves, both has the worst of both repetition and having to look up how the other function works.
In other words, we should either leave this hard-coded and relatively simple, or generalize random_entry to handle weighting.  I'm leaning toward the first, since the second is liable to be hard to generalize.

Comment (kevingranade) [inline src/creature.cpp]: To put it another way, since we're choosing a weight up front, iterating through the list in a random order doesn't make things any better.

**Comment by kevingranade** [comment]
You probably got similar numbers with the debug statements, but here's what I see with a unit test I'm working on for this:
with hit roll 0, all attacks hit torso, I don't think this is intended.

with hit_roll = 1
Torso: 31007
Head: 11383
Eyes: 5803
Left arm: 15725
Right arm: 15809
Left leg: 10152
Right leg: 10121

with hit_roll = 100
Torso: 28897
Head: 20733
Eyes: 10610
Left arm: 11660
Right arm: 11399
Left leg: 8398
Right leg: 8303

In other words, across the full range of normal attack scale, rate of torso hits barely changes, head and eye shots double, and arm and leg hits drop by about 30%.  That doesn't seem like remotely enough variation and head/eyes/torso are way too high overall.  You've addressed the second problem, but not the first.  I'll incorporate your numbers and re-test.

**Comment by kevingranade** [comment]
After applying your constants:
with hit_roll = 1
Torso: 32963
Head: 2346
Eyes: 355
Left arm: 20257
Right arm: 20153
Left leg: 12120
Right leg: 11906

with hit_roll = 100
Torso: 35998
Head: 5098
Eyes: 749
Left arm: 17486
Right arm: 17229
Left leg: 11598
Right leg: 11842

As expected, that's a lot more like we expect, but the change across the range of skills is pretty slim, in fact it's even worse since the magnitude of change in the head and eye area has decreased dramatically, which is where a lot of losses in the other areas were coming from.

**Comment by macrosblackd** [comment]
I can increase the scale of head/eyes based on the hit_roll.

**Comment by macrosblackd** [comment]
Testing without the random_entry portion shows a very high selection of larm/rarm, which isn't what I would be expecting.

File: src/creature.cpp [hunk 3]
```diff
@@ -19,34 +19,34 @@
...
         }
     },
     {
-        1, /* attacker larger */
-        {   { bp_eyes, 5.f },
-            { bp_head, 25.f },
-            { bp_torso, 55.f },
-            { bp_arm_l, 28.f },
-            { bp_arm_r, 28.f },
+        1, /* attacker larger (175pts)*/
```
Comment (i2amroy) [inline src/creature.cpp]: If we're using floats anyways why don't we just convert everything over to a flat 100 point scale with the required decimal values? Using different scales on different things just obfuscates the actual percentage values of the numbers; by having everything on a 100 point scale a developer can easily glance at it and see all the percentages at once for each type, rather than needing to pull out their calculator and divide all of the point values in that category by 1.75, all of the ones in that category by 1.5, and so on.

**Comment by kevingranade** [comment]
I saw that too, I poked at something locally and it went away...
My changes are on #13014
Not trying to supplant your changes, just some parallel tweaks.

**Comment by macrosblackd** [comment]
I've merged changes from #13014 and tweaked the expected weights to match my changes.

File: tests/creature_test.cpp [hunk 4]
```diff
@@ -20,6 +20,16 @@ void calculate_bodypart_distribution( monster &attacker, monster &defender, int
         selected_part_histogram[defender.select_body_part(&attacker, hit_roll)]++;
     }
 
+    if( hit_roll != 0 ) {
```
Comment (kevingranade) [inline tests/creature_test.cpp]: Instead of copying the game formula, I'd rather generate a table of distributions based on a handful of hit rolls (0, 1, 100, maybe a few in between), it nicely illustrates what the goal is by showing what the distributions are.

**Comment by macrosblackd** [comment]
I've also dealt with the pain of merging my changes on top of kevingrande's from #13014
